### PR TITLE
Add a semaphore to NuGet and Mono restore

### PR DESCRIPTION
**Pull request by agocke**
Right now we parallelize by sticking multiple mac builds on the same machine. Since NuGet and Mono toolset restores are machine-global, sometimes the builds stomp on one another. This adds a semaphore to only allow one build to restore at a time.

### CODE REVIEW

**Comment by agocke** [comment]
@jaredpar @jasonmalinowski @tannergooding  + anyone else who wants to review and knows Unix.

**Comment by agocke** [comment]
I'm not totally happy with this implementation. Aside from the fact that we need to remember to call release_sem at every exit point in the restore function, there's no trap for SIGINT/SIGTERM to prevent an interrupt from taking down the script without releasing the mutex.

**Comment by jasonmalinowski** [comment]
>  Aside from the fact that we need to remember to call release_sem at every exit point in the restore function

Can we just use some scripting language that was invented after we were both born? :smile:

File: cibuild.sh [hunk 1]
```diff
@@ -193,10 +222,12 @@ install_mono_toolset()
     tar -jxf $1.tar.bz2
     if [ $? -ne 0 ]; then
         echo "Unable to download toolset"
+        release_sem "$1"
         exit 1
     fi
 
     popd
+    release_sem "$1"
```
Comment (jaredpar) [inline cibuild.sh]: Why do we need it for building?

**Comment by jaredpar** [comment]
I don't think this quite fixes the problem.  Imagine the scenario where we have two changes which are using a different set of NuGet packages.  The only way to safely run such a change is to completely block the semaphore during restore and build.  Correct?

File: cibuild.sh (hunk 1 above)
Comment (agocke) [inline cibuild.sh]: Don't be fooled by the terrible diff - that's in install_mono_toolset()

**Comment by agocke** [comment]
@jaredpar You should only need a semaphore when restoring packages, not building, and the semaphore name for restoring nuget packages is constant -- it's `restore_nuget`, regardless of the nuget version being used.

**Comment by jaredpar** [comment]
@agocke yes but what happens if I'm restoring two different versions of our nuget packages?  Won't the locking here just mean that one clobbers the other more deterministic ally?

**Comment by agocke** [comment]
@jaredpar We no longer delete the existing packages, we just restore additively, so theoretically you can only add packages to the directory and installing two packages will just install the superset of both.

**Comment by agocke** [comment]
@jaredpar @jasonmalinowski Is that a +1? Or is rewriting this in Perl a serious suggestion?

**Comment by jaredpar** [comment]
:+1: 

Was the use of the word "Perl" a clever trick to provoke me to read this PR?  If so good job.

**Pull request merged**